litgpt/data/text_files.py

```diff
@@ -0,0 +1,161 @@
+                "please provide a valid `--tokenizer_dir` path."
+            )
+
+        if not Path(self.out_path_train).is_dir():
+            optimize(
+                fn=partial(tokenize, tokenizer=self.tokenizer),
+                inputs=train_files,
+                output_dir=str(self.out_path_train),
+                num_workers=num_workers,
+                chunk_bytes="50MB",
+            )
+        if not Path(self.out_path_val).is_dir():
+            optimize(
+                fn=partial(tokenize, tokenizer=self.tokenizer),
+                inputs=val_files,
+                output_dir=str(self.out_path_val),
+                num_workers=1,  # there's only 1 file
+                chunk_bytes="50MB",
+            )
+
+    def train_dataloader(self) -> DataLoader:
+        from litdata.streaming import StreamingDataLoader, StreamingDataset, TokensLoader
+
+        train_dataset = StreamingDataset(
+            input_dir=str(self.out_path_train),
+            item_loader=TokensLoader(block_size=self.max_seq_length),
+            shuffle=True,
+            drop_last=True,
+        )
+        if self.num_workers is None:
+            num_workers = get_half_workers()
+
+        train_dataloader = StreamingDataLoader(
+            train_dataset, batch_size=self.batch_size, pin_memory=True, num_workers=num_workers, drop_last=True
+        )
+        return train_dataloader
+
+    def val_dataloader(self) -> DataLoader:
+        from litdata.streaming import StreamingDataset, TokensLoader
+
+        if self.num_workers is None:
+            num_workers = get_half_workers()
+
+        val_dataset = StreamingDataset(
+            input_dir=str(self.out_path_val),
+            item_loader=TokensLoader(block_size=self.max_seq_length),
+            shuffle=True,
+            # Consider setting to False, but we would lose some samples due to truncation when world size > 1
+            drop_last=True,
+        )
+        val_dataloader = DataLoader(
+            val_dataset, batch_size=self.batch_size, pin_memory=True, num_workers=num_workers, drop_last=True
+        )
+        return val_dataloader
+
+
+def tokenize(filename: str, tokenizer: Tokenizer):
+    with open(filename, "r", encoding="utf-8") as file:
+        text = file.read()
+    text = text.strip()
+
+    global_rank = int(os.environ["DATA_OPTIMIZER_GLOBAL_RANK"])
+    num_workers = int(os.environ["DATA_OPTIMIZER_NUM_WORKERS"])
+
+    chunks = []
+    total_length = len(text)
+    num_chunks = max(1, num_workers - 1)
+    chunk_size = total_length // num_chunks
+    for i in range(num_chunks):
+        start_index = i * chunk_size
+        end_index = (i + 1) * chunk_size if i < 9 else total_length
+        chunks.append(text[start_index:end_index])
+
+    local_rank = global_rank % num_workers
+    for example in tqdm(chunks, position=local_rank):
+        tokens = tokenizer.encode(example)
+        yield tokens
```

Comment:
Why is this in a seperate loop instead of yielding directly from the loop above? Why does it need to append to `chunks = []`?

Reply:
Good point, that looks inefficient. The problem is tqdm here which needs an iterable. I updated it so that it doesn't store the text chunks in a list (which is expensive) but rather just the indices. Is this a fair compromise?